File: leetcode-submissions/1019_Next_Greater_Node_In_Linked_List.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function nextLargerNodes(head: ListNode | null): number[] {
    // we need to use the stack to solve this question 
    const stack:number[]=[];
    let vals :number[]=[];
    let ptr:ListNode | null =head;
    while(ptr){
        vals.push(ptr.val);
        ptr=ptr.next;

    }
    const result:number[]=Array.from({length:vals.length},()=>0);
    for(let i:number =0;i<vals.length;i++){
        while(stack.length>0 && vals[stack[stack.length-1]]<vals[i]){
         
                let index :number =stack.pop()!;
                result[index]=vals[i];

              

          
        }
            stack.push(i);
        
    }
   return result;
};

File: leetcode-submissions/1019_Next_Greater_Node_In_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<int> nextLargerNodes(ListNode* head) {
        vector<int>lists;
        ListNode* temp = head;
        while(temp){
            lists.push_back(temp->val);
            temp= temp->next;
        }
        stack<int>st;
        vector<int>ans(lists.size(),0);
        for(int i=0;i<lists.size();i++){
            while(not st.empty() and lists[st.top()]<=lists[i]){
                ans[st.top()]= lists[i];
                st.pop();
            }
            st.push(i);
        }
        return ans ;
    }
};